source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -99,29 +116,45 @@ Http::FilterHeadersStatus DecompressorFilter::encodeHeaders(Http::ResponseHeader
                              *encoder_callbacks_, headers);
 }
 
-Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool) {
-  return maybeDecompress(config_->responseDirectionConfig(), response_decompressor_,
-                         *encoder_callbacks_, data);
+Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (response_decompressor_) {
+    HeaderMapOptRef trailers;
+    if (end_stream) {
+      trailers = HeaderMapOptRef(std::ref(encoder_callbacks_->addEncodedTrailers()));
+    }
+    decompress(config_->responseDirectionConfig(), response_decompressor_, *encoder_callbacks_,
+               data, response_byte_tracker_, trailers);
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus DecompressorFilter::encodeTrailers(Http::ResponseTrailerMap& trailers) {
+  response_byte_tracker_.reportTotalBytes(trailers);
+  return Http::FilterTrailersStatus::Continue;
 }
 
-Http::FilterDataStatus DecompressorFilter::maybeDecompress(
+void DecompressorFilter::decompress(
     const DecompressorFilterConfig::DirectionConfig& direction_config,
     const Compression::Decompressor::DecompressorPtr& decompressor,
-    Http::StreamFilterCallbacks& callbacks, Buffer::Instance& input_buffer) const {
-  if (decompressor) {
-    Buffer::OwnedImpl output_buffer;
-    decompressor->decompress(input_buffer, output_buffer);
-
-    // Report decompression via stats and logging before modifying the input buffer.
-    direction_config.stats().total_compressed_bytes_.add(input_buffer.length());
-    direction_config.stats().total_uncompressed_bytes_.add(output_buffer.length());
-    ENVOY_STREAM_LOG(debug, "{} data decompressed from {} bytes to {} bytes", callbacks,
-                     direction_config.logString(), input_buffer.length(), output_buffer.length());
-
-    input_buffer.drain(input_buffer.length());
-    input_buffer.add(output_buffer);
+    Http::StreamFilterCallbacks& callbacks, Buffer::Instance& input_buffer,
+    ByteTracker& byte_tracker, HeaderMapOptRef trailers) const {
+  ASSERT(decompressor);
+  Buffer::OwnedImpl output_buffer;
+  decompressor->decompress(input_buffer, output_buffer);
+
+  // Report decompression via stats and logging before modifying the input buffer.
+  byte_tracker.chargeBytes(input_buffer.length(), output_buffer.length());
```

Comment:
Checking desired behavior: Do we want to only increment byte counts if we actually do decompression (versus regardless of if we do decompression)?

Reply:
you mean if `decompress` had an issue? Or not doing decompression at all?

Reply:
The latter (i.e., if the response is not compressed)

Reply:
My opinion is that the trailers should only be added if the stream was actually decompressed. So this is correct. But you did make me realize that the decode/encodeTrailers calls were not gated and that was wrong. Fixed in: https://github.com/envoyproxy/envoy/pull/12477/commits/540a5aed80ed29b214bca141ea81090852573b85